Line editing for Data Types document

christinerose opened the pull request:
Stopped at li 111 for now, as the code doesn't work for me there. Awaiting clarification.

christinerose committed:
Line editing for Data Types document

Stopped at li 111 for now, as the code doesn't work for me there. Awaiting clarification.

sabine commented on data/tutorials/lg_00_data_types_and_matching.md:
```diff
@@ -83,6 +83,9 @@ type person = { first_name : string; surname : string; mutable age : int; }
 
+Please note: in the above example, we used the UK spelling of *colour*, but it will also
+work with the US spelling *color*.
+
```
I think emphasizing that either US or UK spelling works is probably less helpful than emphasizing that "colour" is just the name of the type and we could choose any name for the type (though, it helps to choose a name that helps you remember what kind of data the type intends to formalize). Point being that we don't want people to wonder about UK or US english, but instead understand that it's just a name.

```suggestion
Please note: in the above example, we chose "colour" as the name of a type that holds RGB-values. If you wanted to, you could also use the US spelling "color".
```

christinerose committed:
Update data/tutorials/lg_00_data_types_and_matching.md

Co-authored-by: sabine <sabine@users.noreply.github.com>

christinerose committed:
edits through li 430

christinerose committed:
edits through li. 549

This documentation uses "factorizing" instead of the UK spelling "factorising," which is what we usually use. Shall I change this for consistency?

christinerose committed:
added clarification for `#` as command prompt

cuihtlauac commented on data/tutorials/lg_00_data_types_and_matching.md:
```diff
@@ -40,9 +46,9 @@ Next, we have tuples, which collect a fixed number of elements together:
-and their types:
+We have records, which are like labeled tuples. They are sequences, like lists, defined by writing a
+type definition giving a name for the record and names for each of its fields,
+including their types:
```
```suggestion
We have records, which are like labeled tuples. In tuples, a field sits at some place: first, second, third, etc. In records, a field has a name; order is no longer relevant. Therefore, a record definition is a set of field name and type:
```

cuihtlauac reviewed (changes requested):
Only a minor fix.

I need to rewrite a lot of stuff in this document. I'll revisit it soon.

christinerose committed:
Update data/tutorials/lg_00_data_types_and_matching.md

Co-authored-by: Cuihtlauac Alvarado <cuihtlauac@users.noreply.github.com>

christinerose committed:
Suggested edits

sabine commented on data/tutorials/lg_00_data_types_and_matching.md:
```diff
@@ -545,19 +550,11 @@ what are known as *guards* to each pattern match. A guard is the conditional
 
-Another feature which is useful when we build more complicated nested patterns
+Another useful feature when building more complicated nested patterns
 is the `as` keyword, which can be used to name part of an expression. For
```
```suggestion
Another useful feature for building more complicated nested patterns
is the `as` keyword, which can be used to name part of an expression. For
```

sabine commented on data/tutorials/lg_00_data_types_and_matching.md:
```diff
@@ -545,19 +550,11 @@ what are known as *guards* to each pattern match. A guard is the conditional
+Another useful feature when building more complicated nested patterns
 is the `as` keyword, which can be used to name part of an expression. For
 example:
 
```
I would change

<!-- $MDX skip -->
```ocaml
Name ("/DeviceGray" | "/DeviceRGB" | "/DeviceCMYK") as n -> n

Node (l, ((k, _) as pair), r) when k = k' -> Some pair
```

to

<!-- $MDX skip -->
```ocaml
| Name ("/DeviceGray" | "/DeviceRGB" | "/DeviceCMYK") as n -> n

| Node (l, ((k, _) as pair), r) when k = k' -> Some pair
```

as there is a more clear visual reference to these being part of a pattern match.

In this instance, it is fine that it is not directly copyable to a REPL, I think.

However, we need to figure out a good way to reliably link to exercises from within the tutorials, or to have links to pre-filled playgrounds that illustrate the concept in such a way that the reader can tinker with the code. These examples focus on something very specific that isn't fully defined in the text.

christinerose replied:
I changed that as suggested. I like that idea of linking to something where they can play with the code!

sabine reviewed (commented):
The reason for the errors on line 588 is that you need to enter both this block and the previous one into the REPL or into the playground, i.e. in the playground, I would write this:

```
type t' = Int of int | Add of t * t
and t = {annotation : string; data : t'}

let rec sum_t' = function
  | Int i -> i
  | Add (i, i') -> sum_t i + sum_t i'
  and sum_t {annotation; data} =
    if annotation <> "" then Printf.printf "Touching %s\n" annotation;
sum_t' data
```

So the tutorials are picking up pace and are making multiple definitions that depend on previous ones. It would be a good idea to figure out how we can mark such a dependency on a previous block in the documents (so that students can easily see which blocks need to go into the REPL in sequence).

> In li. 622-624, I get the Error message, but I don't get the Line 1, characters 1-7: ... should that be removed? Same with li. 630-633.

Yes, you can remove those line number infos. They are spammy and not every REPL will show them.

> In li. 649, there needs to be an adjective before "operations" to match the first part of the sentence. Otherwise, it reads like it's one single type and a single operation

Suggestion:
"Often, a module will provide a single type, as well as one or more operations on that type."

> Similar question for li 653 - 666 as before. Will the readers know what to do with this information since it apparently isn't meant to be entered into the toplevel on its own.

"For example, a module for a file format like PNG might have the following png.mli interface:"
intends to express that this is an example of a file `png.mli` that contains an interface specification of the module `png.ml`. So indeed it is not meant to be entered into a REPL, but since it is rendered in the same visual styles, it is easy to mix it up with the code blocks meant for REPL pasting. I think we need a general solution to render example code visually different from code lines which are meant to be pasted into a REPL.

christinerose committed:
Updated based on feedback

christinerose committed:
Update data/tutorials/lg_00_data_types_and_matching.md

Co-authored-by: sabine <sabine@users.noreply.github.com>

christinerose committed:
revised code block as suggested
